python/pyspark/util.py

```diff
@@ -406,13 +406,36 @@ def inner(*args: Any, **kwargs: Any) -> Any:
 
         return outer
 
-    # Non Spark Connect
+    # Non Spark Connect with SparkSession or Callable
```

Comment:
@HyukjinKwon I updated the PR to use JVM APIs directly, but wonder if we want to keep the code changes from here to support `InheritableThread` or not

Reply:
Let's document that we should provide `session` to properly inherit the tags. And also, we could issue a warning that the tags wouldn't be inherited when the tags are set but the session is not provided.

Reply:
Sounds good. Just added docs and warnings. Thanks!